pkg/ctl/cmdutils/nodegroup_filter.go

```diff
@@ -72,70 +97,109 @@ func (f *NodeGroupFilter) ApplyExistingFilter(stackManager *manager.StackCollect
 		return err
 	}
 
-	f.existing.Insert(existing...)
+	f.ExcludeNames.Insert(existing...)
 
+	for _, name := range existing {
+		isAlsoIncluded := false
+		f.matchGlobs(name, f.includeGlobs, &isAlsoIncluded)
+		if isAlsoIncluded {
+			return fmt.Errorf("existing nodegroup %q should be excluded, but matches include fileter: %q", name, f.includeGlobsSpec)
+		}
+	}
 	return nil
 }
 
-// Match checks given nodegroup against the filter
-func (f *NodeGroupFilter) Match(ng *api.NodeGroup) bool {
-	if f.SkipAll {
-		return false
+func (*NodeGroupFilter) matchGlobs(name string, exprs []glob.Glob, result *bool) {
+	for _, compiledExpr := range exprs {
+		if compiledExpr.Match(name) {
+			*result = true
+			return
+		}
+	}
+}
+
+// Match given nodegroup against the filter and returns
+// true or false if it has to be included or excluded
+func (f *NodeGroupFilter) Match(name string) bool {
+	if f.ExcludeAll {
+		return false // force exclude
 	}
-	if f.IgnoreAllExisting && f.existing.Has(ng.Name) {
-		return false
+
+	hasIncludeRules := f.IncludeNames.Len()+len(f.includeGlobs) != 0
+	hasExcludeRules := f.ExcludeNames.Len()+len(f.excludeGlobs) != 0
+
+	if !hasIncludeRules && !hasExcludeRules {
+		return true // empty rules - include
+	}
+
+	mustInclude := false // override when rules overlap
+
+	if hasIncludeRules {
+		mustInclude = f.IncludeNames.Has(name)
+		f.matchGlobs(name, f.includeGlobs, &mustInclude)
+		if !hasExcludeRules {
+			// empty exclusion rules - strict inclusion mode
+			return mustInclude
+		}
 	}
 
-	for _, compiledExpr := range f.only {
-		if compiledExpr.Match(ng.Name) {
-			return true // return first match
+	if hasExcludeRules {
+		exclude := f.ExcludeNames.Has(name)
+		f.matchGlobs(name, f.excludeGlobs, &exclude)
+		if exclude && !mustInclude {
+			// no inclusion rules to override
```

Comment:
If the exclude is supposed to override the include, then shouldn't the `if` be like `if exclude { return false }`?

Reply:
Maybe the comment is outdated?

Reply:
Include is supposed to override, via `mustInclude`... comment is somewhat correct, but isn't very clearly phrased.